market/mk12/mk12.go

```diff
@@ -461,3 +499,144 @@ func (m *MK12) processDeal(ctx context.Context, deal *ProviderDealState) (*Provi
+			return
+		}
+	} else {
+		var bufferedSDR, bufferedTrees, bufferedPoRep, waitDealSectors int
+		err = m.db.QueryRow(ctx, `
+		WITH BufferedSDR AS (
+			SELECT COUNT(p.task_id_sdr) - COUNT(t.owner_id) AS buffered_sdr_count
+			FROM sectors_sdr_pipeline p
+			LEFT JOIN harmony_task t ON p.task_id_sdr = t.id
+			WHERE p.after_sdr = false
+		),
+		BufferedTrees AS (
+			SELECT COUNT(p.task_id_tree_r) - COUNT(t.owner_id) AS buffered_trees_count
+			FROM sectors_sdr_pipeline p
+			LEFT JOIN harmony_task t ON p.task_id_tree_r = t.id
+			WHERE p.after_sdr = true AND p.after_tree_r = false
+		),
+		BufferedPoRep AS (
+			SELECT COUNT(p.task_id_porep) - COUNT(t.owner_id) AS buffered_porep_count
+			FROM sectors_sdr_pipeline p
+			LEFT JOIN harmony_task t ON p.task_id_porep = t.id
+			WHERE p.after_tree_r = true AND p.after_porep = false
+		),
+		WaitDealSectors AS (
+			SELECT COUNT(DISTINCT sip.sector_number) AS wait_deal_sectors_count
+			FROM sectors_sdr_initial_pieces sip
+			LEFT JOIN sectors_sdr_pipeline sp ON sip.sp_id = sp.sp_id AND sip.sector_number = sp.sector_number
+			WHERE sp.sector_number IS NULL
+		)
+		SELECT
+			(SELECT buffered_sdr_count FROM BufferedSDR) AS total_buffered_sdr,
+			(SELECT buffered_trees_count FROM BufferedTrees) AS buffered_trees_count,
+			(SELECT buffered_porep_count FROM BufferedPoRep) AS buffered_porep_count,
+			(SELECT wait_deal_sectors_count FROM WaitDealSectors) AS wait_deal_sectors_count
+		`).Scan(&bufferedSDR, &bufferedTrees, &bufferedPoRep, &waitDealSectors)
+		if err != nil {
+			return false, xerrors.Errorf("counting buffered sectors: %w", err)
+		}
+
+		if cfg.MaxQueueDealSector != 0 && waitDealSectors+sectors > cfg.MaxQueueDealSector {
+			log.Infow("backpressure", "reason", "too many wait deal sectors", "wait_deal_sectors", waitDealSectors, "max", cfg.MaxQueueDealSector)
+			return true, nil
+		}
+
+		if bufferedSDR > cfg.MaxQueueSDR {
+			log.Infow("backpressure", "reason", "too many SDR tasks", "buffered", bufferedSDR, "max", cfg.MaxQueueSDR)
+			return true, nil
+		}
+		if cfg.MaxQueueTrees != 0 && bufferedTrees > cfg.MaxQueueTrees {
+			log.Infow("backpressure", "reason", "too many tree tasks", "buffered", bufferedTrees, "max", cfg.MaxQueueTrees)
+			return true, nil
+		}
+		if cfg.MaxQueuePoRep != 0 && bufferedPoRep > cfg.MaxQueuePoRep {
+			log.Infow("backpressure", "reason", "too many PoRep tasks", "buffered", bufferedPoRep, "max", cfg.MaxQueuePoRep)
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func sectorCount(sizes []abi.PaddedPieceSize, targetSize abi.PaddedPieceSize) int {
+	sort.Slice(sizes, func(i, j int) bool {
+		return sizes[i] > sizes[j]
+	})
+
+	sectors := make([]abi.PaddedPieceSize, 0)
+
+	for _, size := range sizes {
+		placed := false
+		for i := range sectors {
+			if sectors[i]+size <= targetSize {
+				sectors[i] += size
+				placed = true
+				break
+			}
+		}
+		if !placed {
+			sectors = append(sectors, size)
+		}
+	}
```

Comment:
Is this a bin-packing algorithm? There are libraries with great guarantees for this NP-Hard problem.

Reply:
In a way it is, but also it's not much of a problem in filecoin-land because the pieces are generally very large, and usually you can fill 100% of the sector with simple greedy strategies

Reply:
This is not required anymore with a more simpler backpressure mechanism.